src/integrationTest/java/org/opensearch/security/rest/GzipTests.java

```diff
@@ -0,0 +1,115 @@
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.zip.GZIPOutputStream;
+
+import static org.junit.Assert.fail;
+import static org.opensearch.test.framework.TestSecurityConfig.AuthcDomain.AUTHC_HTTPBASIC_INTERNAL;
+import static org.opensearch.test.framework.TestSecurityConfig.Role.ALL_ACCESS;
+
+@RunWith(com.carrotsearch.randomizedtesting.RandomizedRunner.class)
+@ThreadLeakScope(ThreadLeakScope.Scope.NONE)
+public class GzipTests {
+    private static final TestSecurityConfig.User ADMIN_USER = new TestSecurityConfig.User("admin").roles(ALL_ACCESS);
+
+    @ClassRule
+    public static LocalCluster cluster = new LocalCluster.Builder().clusterManager(ClusterManager.THREE_CLUSTER_MANAGERS)
+        .authc(AUTHC_HTTPBASIC_INTERNAL)
+        .users(ADMIN_USER)
+        .anonymousAuth(false)
+        .build();
+
+    @Test
+    public void testAuthenticatedGzippedRequests() {
+        final String requestPath = "/*/_search";
+        final int parrallelism = 10;
+        final int totalNumberOfRequests = 100;
+
+        final byte[] compressedRequestBody = createCompressedRequestBody();
+        List<TestRestClient> restClients = new ArrayList<>();
+        for (int i = 0; i < 10; i++) {
+            restClients.add(cluster.getRestClient(ADMIN_USER, new BasicHeader("Content-Encoding", "gzip")));
+        }
+        Random rand = new Random();
+        final HttpPost post = new HttpPost(restClients.get(0).getHttpServerUri() + requestPath);
+        post.setEntity(new ByteArrayEntity(compressedRequestBody, ContentType.APPLICATION_JSON));
+
+        final ForkJoinPool forkJoinPool = new ForkJoinPool(parrallelism);
+
+        final List<CompletableFuture<Void>> waitingOn = IntStream.rangeClosed(1, totalNumberOfRequests)
+            .boxed()
+            .map(i -> CompletableFuture.runAsync(() -> {
+                TestRestClient.HttpResponse response = restClients.get(rand.nextInt(10)).executeRequest(post);
+                assertThat(response.getStatusCode(), equalTo(HttpStatus.SC_OK));
+                assertThat(response.getBody(), not(containsString("json_parse_exception")));
+            }, forkJoinPool))
+            .collect(Collectors.toList());
+
+        final CompletableFuture<Void> allOfThem = CompletableFuture.allOf(waitingOn.toArray(new CompletableFuture[0]));
+
+        try {
+            allOfThem.get(30, TimeUnit.SECONDS);
+        } catch (final Exception e) {
+            Throwable rootCause = e.getCause();
+            if (rootCause instanceof AssertionError) {
+                fail("Received exception: " + e.getMessage());
+            }
+        }
```

Comment:
Isn't throwing exception ok? If the root cause isn't an assert exception we will swallow the exception, shouldn't the test fail?
```suggestion
         allOfThem.get(30, TimeUnit.SECONDS);
```

Reply:
I thought so too, but when I remove this I get IllegalStateException: Endpoint not acquired / already released

Reply:
Doesn't that mean some of the requests aren't working correctly?

Reply:
It looks like its coming from [here](https://github.com/apache/httpcomponents-client/blob/19f3922b37369431d379ce57fa187740b22436cf/httpclient5/src/main/java/org/apache/hc/client5/http/impl/async/InternalHttpAsyncExecRuntime.java#L190), but I haven't been able to determine yet where it is thrown.

Reply:
This is the critical feedback to address - we need to have certainty that expected to succeed requests are always working, and that failures are within a range of expectations, such as 401s.  If we need to trigger the requests in a slightly different way to side step this issue I'm fine with that - e.g. the issue can be reproduce by doing 3x 200 requests with a 401 interleaved within each IMO that is good enough for this change.


Note; The resource tests were focused specifically on generating high volumes of requests, not so much on validating them so there might be work to patch up that process.